ds701_book/02B-Pandas-Sklearn.qmd

```diff
@@ -0,0 +1,610 @@
+
+The correlation coefficient between variables $X$ and $Y$ is defined as follows
+
+$$ \text{Corr}(X,Y) = \frac{E\left[(X-\mu_X)(Y-\mu_Y)\right]}{\sigma_X\sigma_Y}. $$
+
+Pandas provides a DataFrame method called `corr()` that computes the correlation coefficient of all pairs of columns.
+
+```{python}
+rets.corr()
+```
+
+It takes a bit of time to examine that table and draw conclusions.  
+
+To speed that process up let's visualize the table.
+
+
+```{python}
+import seaborn as sns
+
+sns.heatmap(rets.corr(), annot=True)
+plt.show()
+```
+
+Finally, it is important to know that the plotting performed by Pandas is just a layer on top of `matplotlib` (i.e., the `plt` package).  
+
+So Panda's plots can (and often should) be replaced or improved by using additional functions from `matplotlib`.
+
+For example, suppose we want to know both the returns as well as the standard deviation of the returns of a stock (i.e., its risk).  
+
+Here is a visualization of the result of such an analysis. We construct the plot using only functions from `matplotlib`.
+
+```{python}
+plt.scatter(rets.mean(), rets.std())
+plt.xlabel('Expected returns')
+plt.ylabel('Standard Deviation (Risk)')
+plt.xlim([-.05,.1])
+plt.ylim([0,.3])
+for label, x, y in zip(rets.columns, rets.mean(), rets.std()):
+    plt.annotate(
+        label, 
+        xy = (x, y), xytext = (30, -30),
+        textcoords = 'offset points', ha = 'right', va = 'bottom',
+        bbox = dict(boxstyle = 'round,pad=0.5', fc = 'yellow', alpha = 0.5),
+        arrowprops = dict(arrowstyle = '->', connectionstyle = 'arc3,rad=0'))
+plt.show()
+```
+
+To understand what these functions are doing, (especially the [annotate](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.annotate.html) function), you will need to consult the online documentation for [matplotlib](https://matplotlib.org/stable/api/index.html). 
+
+# Scikit-Learn
+
+Scikit-Learn (sklearn) is a powerful Python package for machine learning. The theoretical underpinnings of the methods introduced in this lecture will be covered in later lectures. The intent of this lecture is to demonstrate
+
+1. how to implement a few machine learning models using Scikit-Learn and
+1. understand the structure of the Scikit-Learn API.
+
+The 2nd point is key, because understanding the generality of the API, will allow you to easily work with different sklearn machine learning models.
+
+The general framework for implementing a machine learning models in sklearn is
+
+1. Import the sklearn objects you need.
+1. Prepare the dataset.
+1. Instatiate a machine learning object.
+1. Train your model on the dataset.
+1. Use the trained model to make predictions.
+1. Evaluate the performance of the model. 
+
+We will demonstrate the sklearn machine learning framework by working with the sklearn `LinearRegression` object. This object can be used to train a linear model that predicts continuous values. 
```

Comment:
Makes sense to to use an example, but we do have the linear regrression chapter/lecture. Did you reconcile with that? I'm also reviewing now lecture 07 and it is introducing sklearn a bit too.

Reply:
I'm happy to make a separate lecture. I think that will make the essential tools chatper less bulky and easier to read.

I used linear regression because I had used it previously in my RCS tutorials. I referenced that we cover more theoretical/implementation details in the Linear Regression section. 

However, I think I should also look at the introduction that is made in lecture 07 and see if there is anything that could be moved from there to the sklearn chapter and then cross-reference from 07.